x-pack/plugins/ml/public/data_frame/components/source_index_preview/source_index_preview.tsx

```diff
@@ -126,35 +123,65 @@ export const SourceIndexPreview: React.SFC<Props> = React.memo(({ cellClick, que
     setItemIdToExpandedRowMap({ ...itemIdToExpandedRowMap });
   }
 
-  useEffect(
-    () => {
-      setLoading(true);
-
-      ml.esSearch({
-        index: indexPattern.title,
-        rest_total_hits_as_int: true,
-        size: SEARCH_SIZE,
-        body: { query },
-      })
-        .then((resp: SearchResponse<any>) => {
-          const docs = resp.hits.hits;
-
-          if (selectedFields.length === 0) {
-            const newSelectedFields = getDefaultSelectableFields(docs);
-            setSelectedFields(newSelectedFields);
-          }
-
-          setTableItems(docs as EsDoc[]);
-          setLoading(false);
-        })
-        .catch((resp: any) => {
-          setTableItems([] as EsDoc[]);
-          setLoading(false);
-        });
-    },
-    [indexPattern.title, query.query_string.query]
+  const { errorMessage, status, tableItems } = useSourceIndexData(
+    indexPattern,
+    query,
+    selectedFields,
+    setSelectedFields
   );
 
+  if (status === SOURCE_INDEX_STATUS.ERROR) {
+    return (
+      <EuiPanel grow={false}>
+        <SourceIndexPreviewTitle indexPatternTitle={indexPattern.title} />
+        <EuiCallOut
+          title={i18n.translate('xpack.ml.dataframe.sourceIndexPreview.sourceIndexPatternError', {
+            defaultMessage: 'An error occurred loading the source index data.',
+          })}
+          color="danger"
+          iconType="cross"
+        >
+          <p>{errorMessage}</p>
+        </EuiCallOut>
+      </EuiPanel>
+    );
+  }
+
+  if (status === SOURCE_INDEX_STATUS.LOADED && tableItems.length === 0) {
+    return (
+      <EuiPanel grow={false}>
+        <SourceIndexPreviewTitle indexPatternTitle={indexPattern.title} />
+        <EuiCallOut
+          title={i18n.translate(
+            'xpack.ml.dataframe.sourceIndexPreview.dataFrameSourceIndexNoDataCalloutTitle',
+            {
+              defaultMessage: 'Empty source index query result.',
+            }
+          )}
+          color="primary"
+        >
+          <p>
+            {i18n.translate(
+              'xpack.ml.dataframe.sourceIndexPreview.dataFrameSourceIndexNoDataCalloutBody',
+              {
+                defaultMessage:
+                  'The query for the source index returned no results. Please make sure the index contains documents and your query is not too limiting.',
```

Comment:
Maybe change 'limiting' to 'restrictive'?

Reply:
Changed in a5ee6fb.